{"expand":"renderedFields,names,schema,transitions,operations,editmeta,changelog","id":"12314290","self":"https://issues.apache.org/jira/rest/api/latest/issue/12314290","key":"LUCENE-140","fields":{"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/images/icons/issuetypes/bug.png","name":"Bug","subtask":false},"timespent":null,"project":{"self":"https://issues.apache.org/jira/rest/api/2/project/12310110","id":"12310110","key":"LUCENE","name":"Lucene - Core","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/projectavatar?pid=12310110&avatarId=10061","24x24":"https://issues.apache.org/jira/secure/projectavatar?size=small&pid=12310110&avatarId=10061","16x16":"https://issues.apache.org/jira/secure/projectavatar?size=xsmall&pid=12310110&avatarId=10061","32x32":"https://issues.apache.org/jira/secure/projectavatar?size=medium&pid=12310110&avatarId=10061"},"projectCategory":{"self":"https://issues.apache.org/jira/rest/api/2/projectCategory/10150","id":"10150","description":"Lucene-related projects","name":"Lucene"}},"fixVersions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12310854","id":"12310854","description":"","name":"2.1","archived":false,"released":true,"releaseDate":"2007-02-17"}],"aggregatetimespent":null,"resolution":{"self":"https://issues.apache.org/jira/rest/api/2/resolution/1","id":"1","description":"A fix for this issue is checked into the tree and tested.","name":"Fixed"},"customfield_12312322":null,"customfield_12310220":"2003-11-25 19:35:32.0","customfield_12312323":null,"customfield_12310420":"13610","customfield_12312320":null,"customfield_12310222":"1_*:*_1_*:*_102960506300_*|*_6_*:*_1_*:*_0_*|*_5_*:*_1_*:*_4082173739","customfield_12312321":null,"customfield_12312120":null,"customfield_12312121":null,"resolutiondate":"2007-01-11T12:14:18.300+0000","workratio":-1,"customfield_12312328":null,"customfield_12312329":null,"customfield_12312326":null,"customfield_12310300":null,"customfield_12312327":null,"customfield_12312324":null,"customfield_12312325":null,"lastViewed":null,"watches":{"self":"https://issues.apache.org/jira/rest/api/2/issue/LUCENE-140/watchers","watchCount":5,"isWatching":false},"created":"2003-10-07T20:05:52.000+0000","priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.png","name":"Major","id":"3"},"labels":[],"customfield_12312333":null,"customfield_12312334":null,"customfield_12310310":"5.0","customfield_12312331":null,"customfield_12312332":null,"timeestimate":null,"aggregatetimeoriginalestimate":null,"customfield_12312330":null,"customfield_12311120":null,"versions":[],"issuelinks":[],"customfield_12312339":null,"assignee":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"customfield_12312337":null,"customfield_12312338":null,"updated":"2011-06-02T22:04:35.962+0000","customfield_12312335":null,"customfield_12312336":null,"customfield_12311720":null,"status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Complete"}},"components":[{"self":"https://issues.apache.org/jira/rest/api/2/component/12310232","id":"12310232","name":"core/index","description":"issues with indexing code"}],"timeoriginalestimate":null,"description":"Hello,\n  I can not find out, why (and what) it is happening all the time. I got an\nexception:\njava.lang.IllegalStateException: docs out of order\n        at\norg.apache.lucene.index.SegmentMerger.appendPostings(SegmentMerger.java:219)\n        at\norg.apache.lucene.index.SegmentMerger.mergeTermInfo(SegmentMerger.java:191)\n        at\norg.apache.lucene.index.SegmentMerger.mergeTermInfos(SegmentMerger.java:172)\n        at org.apache.lucene.index.SegmentMerger.mergeTerms(SegmentMerger.java:135)\n        at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:88)\n        at org.apache.lucene.index.IndexWriter.mergeSegments(IndexWriter.java:341)\n        at org.apache.lucene.index.IndexWriter.optimize(IndexWriter.java:250)\n        at Optimize.main(Optimize.java:29)\n\nIt happens either in 1.2 and 1.3rc1 (anyway what happened to it? I can not find\nit neither in download nor in version list in this form). Everything seems OK. I\ncan search through index, but I can not optimize it. Even worse after this\nexception every time I add new documents and close IndexWriter new segments is\ncreated! I think it has all documents added before, because of its size.\n\nMy index is quite big: 500.000 docs, about 5gb of index directory.\n\nIt is _repeatable_. I drop index, reindex everything. Afterwards I add a few\ndocs, try to optimize and receive above exception.\n\nMy documents' structure is:\n  static Document indexIt(String id_strony, Reader reader, String data_wydania,\nString id_wydania, String id_gazety, String data_wstawienia)\n{\n    Document doc = new Document();\n\n    doc.add(Field.Keyword(\"id\", id_strony ));\n    doc.add(Field.Keyword(\"data_wydania\", data_wydania));\n    doc.add(Field.Keyword(\"id_wydania\", id_wydania));\n    doc.add(Field.Text(\"id_gazety\", id_gazety));\n    doc.add(Field.Keyword(\"data_wstawienia\", data_wstawienia));\n    doc.add(Field.Text(\"tresc\", reader));\n\n    return doc;\n}\n\nSincerely,\nlegez","customfield_10010":23650.0,"timetracking":{},"customfield_12310120":null,"customfield_12312026":null,"customfield_12312023":null,"customfield_12312024":null,"customfield_12312340":null,"aggregatetimeestimate":null,"customfield_12312022":null,"customfield_12312341":null,"customfield_12312220":null,"customfield_12310921":null,"customfield_12310920":"27909","summary":"docs out of order","creator":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=gegez%40wp.pl","name":"gegez@wp.pl","emailAddress":"gegez at wp dot pl","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"legez","active":true},"subtasks":[],"customfield_12310291":null,"reporter":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=gegez%40wp.pl","name":"gegez@wp.pl","emailAddress":"gegez at wp dot pl","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"legez","active":true},"customfield_12310290":null,"aggregateprogress":{"progress":0,"total":0},"customfield_12311024":null,"environment":"Operating System: Linux\nPlatform: PC","customfield_12311020":null,"duedate":null,"customfield_12310250":null,"progress":{"progress":0,"total":0},"comment":{"startAt":0,"maxResults":37,"total":37,"comments":[{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12321387","id":"12321387","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=otis%40apache.org","name":"otis@apache.org","emailAddress":"otis at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Otis Gospodnetic","active":true},"body":"Since the bug submitter hasn't followed up to this 'repeatable' issue in almost\r\ntwo months, I shall assume that this was not a bug in Lucene, but a misuse of\r\nLucene.\r\n\r\nThere is also no self-contained code that demonstrates the bug.\r\n\r\nIf this is indeed a Lucene bug, please re-open this bug entry, and provise\r\nself-sufficient unit test that demonstrates this problem.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=otis%40apache.org","name":"otis@apache.org","emailAddress":"otis at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Otis Gospodnetic","active":true},"created":"2003-11-25T19:35:32.000+0000","updated":"2003-11-25T19:35:32.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12321388","id":"12321388","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=lucene%40ziplip.com","name":"lucene@ziplip.com","emailAddress":"lucene at ziplip dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Arvind Srinivasan","active":true},"body":"More Data Integrity Issue: Docs out of Order\r\n\r\nHi,\r\nSeeing similar issue to the one reported in:\r\nhttp://nagoya.apache.org/bugzilla/show_bug.cgi?id=23650\r\nOn examining the segments, following inconsistencies were found\r\n(a) The merging segments had doc number that is greater than maxDoc.\r\nDon't know how it go in this state, but this occurs using standard lucene\r\ncode.\r\n(b) Strangely, some documents had terms with zero frequency.  And when it \r\noccurred,\r\nthe zero frequency term has several posting as (docid 0)\r\nExample.. (docid freq)  -- MaxDoc = 7749 - NO DELETION.\r\nMerging msgBody; text=it; sz=2  --- The field name is msgBody and term is \"it\"\r\n                                    and two segments have the term.\r\n(0 0)(0 0)(0 0)..........(0 0)(4 6)(5 3)(6 1)(9 1)(10 2)(12 1)......\r\n...(6791 2)(6794 3)(6796 2)(6798 16)(6801 1)(6805 1)(6806 5)\r\n(6808 1)(6810 1)(6815 2)(6816 3)(6817 1)(6818 1)(6821 4)(6822 1)\r\n(6824 3)(6826 4)(6828 1)(6829 3)(12549 2)doc exceeds count\r\n749(13570 1)doc exceeds count 7749(14896 1)doc exceeds count 7749\r\n(15028 1)doc exceeds count 7749(15357 1)doc exceeds count 7749\r\n(15427 1)doc exceeds count 7749(15534 1)doc exceeds count 7749\r\n(15535 1)doc exceeds count 7749(15653 1)doc exceeds count 7749\r\n(16530 1)doc exceeds count 7749(17108 1).......\r\n(c) Also the zero frequency was not limited to the 0 document, there was\r\nanother instance.\r\n\r\nOne work around that seemed to resolve the issue was to:\r\n(a) keep the maxDoc as a member variable in SegmentMergeInfo\r\nand ignore/throw exception if an inconsistent state is detected.\r\n\r\n****ADD To SegmentMerger just before \"docs out of order\" check.\r\n  if (postings.freq() == 0) {\r\n            continue;\r\n   }\r\n   if (doc >= smi.maxDoc) {\r\n      //sbLog.append(\"doc exceeds count \\r\\n \" + smi.maxDoc);\r\n      continue;\r\n   }\r\n****\r\n\r\nAtleast putting a check would not corrupt the segments and would\r\nget us closer to the real problem as to why freq=0 and doc number exceeds\r\nmaxDoc. Note, the code has had the fix to the other Segment corruption issue\r\nthat I previously reported (namely, Using a zero length file).\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=lucene%40ziplip.com","name":"lucene@ziplip.com","emailAddress":"lucene at ziplip dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Arvind Srinivasan","active":true},"created":"2005-06-14T23:42:25.000+0000","updated":"2005-06-14T23:42:25.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12321389","id":"12321389","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=lucene%40ziplip.com","name":"lucene@ziplip.com","emailAddress":"lucene at ziplip dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Arvind Srinivasan","active":true},"body":"Created an attachment (id=15405)\r\nAnalysis of corrupted segments and suggestions.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=lucene%40ziplip.com","name":"lucene@ziplip.com","emailAddress":"lucene at ziplip dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Arvind Srinivasan","active":true},"created":"2005-06-14T23:46:42.000+0000","updated":"2005-06-14T23:46:42.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12321390","id":"12321390","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=otis%40apache.org","name":"otis@apache.org","emailAddress":"otis at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Otis Gospodnetic","active":true},"body":"Arvind, thanks for following up on this issue.  From your report I can see that\r\nthe index really ends up containing invalid data, but I've never even seen this\r\nhappen myself.  Could you please put together sample code that gets the index in\r\nthis state?","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=otis%40apache.org","name":"otis@apache.org","emailAddress":"otis at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Otis Gospodnetic","active":true},"created":"2005-06-22T12:04:49.000+0000","updated":"2005-06-22T12:04:49.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12357178","id":"12357178","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=yseeley%40gmail.com","name":"yseeley@gmail.com","emailAddress":"yseeley at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Yonik Seeley","active":true},"body":"I've never seen this... can anyone reproduce with Lucene 1.9?\r\nCCing this to Arvind's email...","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=yseeley%40gmail.com","name":"yseeley@gmail.com","emailAddress":"yseeley at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Yonik Seeley","active":true},"created":"2005-11-10T11:59:30.000+0000","updated":"2005-11-10T11:59:30.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12357180","id":"12357180","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=otis","name":"otis","emailAddress":"otis at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=otis&avatarId=15859","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=otis&avatarId=15859","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=otis&avatarId=15859","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=otis&avatarId=15859"},"displayName":"Otis Gospodnetic","active":true},"body":"2 years later, I still haven't seen this error.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=otis","name":"otis","emailAddress":"otis at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=otis&avatarId=15859","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=otis&avatarId=15859","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=otis&avatarId=15859","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=otis&avatarId=15859"},"displayName":"Otis Gospodnetic","active":true},"created":"2005-11-10T12:05:56.000+0000","updated":"2005-11-10T12:05:56.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12362310","id":"12362310","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"body":"First I think that Lucene is great and it manages to do an incredible job. However, this issue is also causing us significant problems. We originally had an AOP interceptor that would update documents in our HTTP threads however when we started to see this issue we were concerned that it was caused by multiple threads accessing the index at the same time. We put extra concurrency controls on the updates using the LuceneIndexAccess API that was posted into bugzilla by another user. This issue still remained after we added the extra concurrency control.\r\n\r\nSince then we have abandoned the AOP approach completely and moved the index rebuild to a scheduled thread which collects things that were modified from the database (using datestamp versioning) and then proceeds to update their lucene indexes. We hoped this would solve the problem because only 1 thread in 1 process would be modifying the index at any given time. Alas, we are still getting docs out of order exceptions. It is difficult for us to reproduce as it mainly happens in production and we cannot provide a testcase for it (I wish we could!). \r\n\r\nI know that this must be a difficult issue because it is difficult to reproduce. I can't think of anything strange that we are doing with the indexes (one process, one thread modifying). This probably isn't much help but I just wanted to let you know that we are also experiencing the problem. ","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"created":"2006-01-10T16:30:44.000+0000","updated":"2006-01-10T16:30:44.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12362354","id":"12362354","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=cutting","name":"cutting","emailAddress":"cutting at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Doug Cutting","active":true},"body":"File corruption could cause this.  Please look in your system logs to see if there are any reports of problems accessing the drive that stores this index.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=cutting","name":"cutting","emailAddress":"cutting at apache dot org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Doug Cutting","active":true},"created":"2006-01-11T02:28:02.000+0000","updated":"2006-01-11T02:28:02.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12362866","id":"12362866","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"body":"Hi Doug,\r\n\r\nThanks for your suggestion (and great work on Lucene!). I looked in the logs and could not find any indication of corruption. In addition we actually have the system running in a cluster where each node in the cluster has its own lucene search index. We had the issue on a different machine before we moved to the cluster and we now see it on both machines in our cluster :(. Next time it occurs I will get the indexes and try to post them here if you think that would be helpful.\r\n\r\nJust some additional info: \r\n1)  We are running SUSE 10 Linux.\r\n2)  We run two quartz jobs: One job runs every 2 minutes and updates lucene with changes from the db if necessary. The other job runs at 3:30AM in the morning and does full index rebuilds. We use the LuceneIndexAccess api when working with the IndexReader/IndexWriter. Only one thread should modify the index at any time although it is possible that the full rebuild job and the update job execute concurrently they shouldn't both modify the index due to the LuceneIndexAccess API's synchronization.\r\n3) Currently when doing searches we create a new IndexSearcher everytime a search is to be performed; We do not use the LuceneIndexAccess API.\r\n\r\nWhen I refer to the LuceneIndexAccess API I am referring to the contribution by Mark Schreiber:\r\nhttp://www.blizzy.de/lucene/lucene-indexaccess-0.1.0.zip\r\n\r\nThanks for any help!\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"created":"2006-01-17T02:45:00.000+0000","updated":"2006-01-17T02:45:00.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12363156","id":"12363156","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=rmajewski","name":"rmajewski","emailAddress":"rmajewski at infovide dot pl","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Rafal Majewski","active":true},"body":"Hi,\r\n\r\nI have recently experienced the same problem on relase 1.4.3. It happended in production(more than once). Filesystem index directory is used. One application is accountable for indexing, another for searching the index(it also uses IndexReader for query rewriting). Access to index directory(which is singleton) is synchronized within each application's scope. These two applications create separate directory instances and access it independently. Unfortunately,  the code doesn't lead to repeatable occurances of this exception. I assume that  these two applications having not synchronized access to the index might couse the problem, but I have received information that it had also occured with the searching application being down.\r\n\r\nThanks in advance for any help.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=rmajewski","name":"rmajewski","emailAddress":"rmajewski at infovide dot pl","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Rafal Majewski","active":true},"created":"2006-01-19T05:57:46.000+0000","updated":"2006-01-19T05:57:46.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12363471","id":"12363471","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"body":"I am posting our corrupted index (I have to do it in two  parts because it is 14.5M). I looked at it in Luke but Luke doesn't really have any tools to help really diagnose corruption issues. At this point we are considering changing our system to do builds on one machine, test them, and then have them distributed to the other machines in our cluster.\r\n\r\nIf anybody could look at this it would be greatly appreciated!\r\nThanks!\r\nJarrod","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"created":"2006-01-21T07:41:44.000+0000","updated":"2006-01-21T07:41:44.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12363473","id":"12363473","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"body":"Second part. :)","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jcuzens","name":"jcuzens","emailAddress":"maxjar10 at yahoo dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jarrod Cuzens","active":true},"created":"2006-01-21T07:44:02.000+0000","updated":"2006-01-21T07:44:02.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12374312","id":"12374312","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=alesj","name":"alesj","emailAddress":"ales dot justin at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ales Justin","active":true},"body":"We used Lucene 1.4.3. for a half year now.\r\nBut just out of the sudden this issue appeared.\r\n\r\nWe have synchronized access to index - synchronized singleton handling all write / read operations.\r\n\r\nIs there any progress on finding the cause?\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=alesj","name":"alesj","emailAddress":"ales dot justin at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ales Justin","active":true},"created":"2006-04-13T16:13:41.000+0000","updated":"2006-04-13T16:13:41.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12376780","id":"12376780","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jlambert","name":"jlambert","emailAddress":"jlambert at istop dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jason Lambert","active":true},"body":"I was having this problem intermittently while indexing over multiple threads and I have found that the following steps can cause this error (with Lucene 1.3 and 1.4.x):\r\n\r\n- Open an IndexReader (#1) over an existing index (this reader is used for searching while updating the index)\r\n- Using this reader (#1) do a search for the document(s) that you would like to update; obtain their document ID numbers\r\n- Create an IndexWriter and add several new documents to the index (for me, this writing is done in other threads) (*)\r\n- Close the IndexWriter (*)\r\n- Open another IndexReader (#2) over the index\r\n- Delete the previously found documents by their document ID numbers using reader #2\r\n- Close the #2 reader\r\n- Create another IndexWriter (#2) and re-add the updated documents\r\n- Close the IndexWriter #2\r\n- Close the original IndexReader (#1) and open a new reader for general searching\r\n \r\nIf I ensure that the steps marked with an asterisk (*) do not happen (i.e. using thread synchronization), I never get this error.  Otherwise, it will happen intermittently while closing the second IndexWriter (#2)  (sometimes I get an ArrayIndexOutOfBoundsException during the deletion).  These 'extra' writes cause the initial 'segments' file to be updated after which it is re-read while opening the second IndexReader (#2).\r\n\r\nThis can end up deleting documents using possibly non-existent IDs, most likely causing the index corruption that this error signals.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jlambert","name":"jlambert","emailAddress":"jlambert at istop dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jason Lambert","active":true},"created":"2006-04-28T01:49:21.000+0000","updated":"2006-04-28T01:49:21.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12458669","id":"12458669","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"We have seen this one as well. We don't have the same usage as above, we only ever delete documents with IndexReader.deleteDocuments(Term)\r\n\r\nWe are using Lucene 1.9.1\r\n\r\nIt occurs in two places, inside IndexWriter.addDocument():\r\n\r\njava.lang.IllegalStateException: docs out of order\r\n\tat org.apache.lucene.index.SegmentMerger.appendPostings([Lorg/apache/lucene/index/SegmentMergeInfo;I)I(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.mergeTermInfo([Lorg/apache/lucene/index/SegmentMergeInfo;I)V(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.mergeTermInfos()V(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.mergeTerms()V(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.merge()I(Optimized Method)\r\n\tat org.apache.lucene.index.IndexWriter.mergeSegments(II)V(IndexWriter.java:681)\r\n\tat org.apache.lucene.index.IndexWriter.mergeSegments(I)V(IndexWriter.java:658)\r\n\tat org.apache.lucene.index.IndexWriter.maybeMergeSegments()V(IndexWriter.java:646)\r\n\tat org.apache.lucene.index.IndexWriter.addDocument(Lorg/apache/lucene/document/Document;Lorg/apache/lucene/analysis/Analyzer;)V(IndexWriter.java:453)\r\n\tat org.apache.lucene.index.IndexWriter.addDocument(Lorg/apache/lucene/document/Document;)V(IndexWriter.java:436)\r\n\r\nand inside IndexWriter.close():\r\n\r\njava.lang.IllegalStateException: docs out of order\r\n\tat org.apache.lucene.index.SegmentMerger.appendPostings([Lorg/apache/lucene/index/SegmentMergeInfo;I)I(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.mergeTermInfo([Lorg/apache/lucene/index/SegmentMergeInfo;I)V(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.mergeTermInfos()V(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.mergeTerms()V(Optimized Method)\r\n\tat org.apache.lucene.index.SegmentMerger.merge()I(Optimized Method)\r\n\tat org.apache.lucene.index.IndexWriter.mergeSegments(II)V(IndexWriter.java:681)\r\n\tat org.apache.lucene.index.IndexWriter.mergeSegments(I)V(IndexWriter.java:658)\r\n\tat org.apache.lucene.index.IndexWriter.flushRamSegments()V(IndexWriter.java:628)\r\n\tat org.apache.lucene.index.IndexWriter.close()V(IndexWriter.java:375)\r\n\r\nThe second one exposes a problem in the close() method which is that the index write.lock is not released when exceptions are thrown in close() causing subsequent attempts to open an IndexWriter to fail.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2006-12-15T04:11:17.000+0000","updated":"2006-12-15T04:11:17.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12459457","id":"12459457","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"I just resolved LUCENE-603 as a dup of this issue.\r\n\r\nIt would be awesome if we could get a test case that shows this happening.  Enough people seem to hit it that it seems likely something is lurking out there so I'd like to get it fixed!!","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2006-12-18T20:57:43.000+0000","updated":"2006-12-18T20:57:43.000+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12462949","id":"12462949","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"We have now seen this in a number of customer sites since upgrading JIRA to use Lucene 1.9.1. The JIRA report is here: http://jira.atlassian.com/browse/JRA-11861\r\n\r\nWe only seem to have seen it since the upgrade from 1.4.3 to 1.9.1, we hadn't seen it before then.\r\n\r\nThis is now a major issue for us, it is hitting a number of our customers. I am trying to generate a repeatable test for it as a matter of urgency.\r\n\r\nAs a follow-up we sometimes see the old ArrayIndexOutOfBoundsEx in BitVector.get() (BitVector.java:63)\r\n\r\nwill post more if I find something worth sharing.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-08T05:42:15.803+0000","updated":"2007-01-08T05:42:15.803+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12462950","id":"12462950","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"and we also see ArrayIndexOutOfBoundsEx in the SegmentReader.isDeleted() method:\r\n\r\njava.lang.ArrayIndexOutOfBoundsException\r\n        at org.apache.lucene.index.SegmentReader.isDeleted(I)Z(Optimized Method)\r\n        at org.apache.lucene.index.SegmentMerger.mergeFields()I(Optimized Method)\r\n        at org.apache.lucene.index.SegmentMerger.merge()I(Optimized Method)\r\n        at org.apache.lucene.index.IndexWriter.mergeSegments(II)V(IndexWriter.java:681)","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-08T05:44:17.130+0000","updated":"2007-01-08T05:44:17.130+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463029","id":"12463029","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"OK: I finally found one way that this corruption can occur!  I will\r\ncreate a unit test & commit a fix.\r\n\r\nIf you delete by document number, and, that document number is larger\r\nthan maxDoc, but only by a little, then the call to\r\ndeletedDocs.set(num) may in fact succeed (ie, no exception), but will\r\nhave set bits that are \"out of bounds\" in the BitVector's bits array.\r\n\r\nThis is because the bits array is an array of bytes and so you can\r\nhave up to 7 of these unused bits at the end.  Once this has happened,\r\nany attempt to merge this segment will hit the \"docs out of order\"\r\nexception because the BitVector's count() method will count these\r\n\"illegally set\" bits and thus make the SegmentMerger think too many\r\ndocs are deleted. \r\n\r\nUnfortunately, this case only occurs if you use deleteDocument(int),\r\nso I can't yet explain how this happens when using only\r\ndeleteDocument(Term).\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-08T13:09:05.453+0000","updated":"2007-01-08T13:09:05.453+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463093","id":"12463093","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"\r\nI've committed a fix for this one case to the trunk.\r\n\r\nI'm leaving the issue open so folks above can try the fix and confirm\r\nwhether or not this fixes their cases.\r\n\r\nJed (or any other folks who have hit this above and are still\r\nlistening!), the fix is really trivial and would be easy to back\r\nport to prior releases: just run \"svn diff -r 494135:494136\" from\r\na Lucene checkout to see them.\r\n\r\nIf you are willing/able to try this in one of the environments where\r\nyou keep hitting this issue, that would be awesome: if this is in fact\r\nyour root cause, then you would see an ArrayIndexOutOfBoundsException\r\nat the point that the delete of a too-large docNum occurred (rather\r\nthan silent corruption and the above exception much later that you now\r\nsee); and if it's not your root cause after testing the fix, then we\r\nwould know for sure to look for another cause here.\r\n\r\nAre you sure that you only ever do IndexReader.deleteDocuments(Term)\r\nand not deleteDocuments(int docNum)?  I still can't explain how this\r\nerror could happen without using that second method.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-08T18:34:10.671+0000","updated":"2007-01-08T18:34:10.671+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463176","id":"12463176","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=doronc","name":"doronc","emailAddress":"cdoronc at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=doronc&avatarId=12738","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=doronc&avatarId=12738","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=doronc&avatarId=12738","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=doronc&avatarId=12738"},"displayName":"Doron Cohen","active":true},"body":"Amazed by this long lasting bug report I was going similar routes to Mike, and I noticed 3 things - \r\n\r\n(1) the sequence of ops brought by Jason is wrong: \r\n -a- Open an IndexReader (#1) over an existing index (this reader is used for searching while updating the index)\r\n -b- Using this reader (#1) do a search for the document(s) that you would like to update; obtain their document ID numbers\r\n -c- Create an IndexWriter and add several new documents to the index (for me, this writing is done in other threads) (*)\r\n -d- Close the IndexWriter (*)\r\n -e- Open another IndexReader (#2) over the index\r\n -f- Delete the previously found documents by their document ID numbers using reader #2\r\n -g- Close the #2 reader\r\n -h- Create another IndexWriter (#2) and re-add the updated documents\r\n -i- Close the IndexWriter #2\r\n -j- Close the original IndexReader (#1) and open a new reader for general searching\r\n\r\nProblem here is that the docIDs found in (b) may be altered in step (d) and so step (f) would delete the wrong docs. In particular, it might attempt to delete ids that are out of the range. This might expose exactly the BitVector problem, and would explain the whole thing, but I too cannot see how it explains the delete-by-term case.\r\n\r\n(2) BitVectort silent ignoring of attempts to delete slightly-out-of-bound docs that fall in the higher byte - this the problem that Mike fixed. I think the fix is okay - though some applications might now get exceptions they did not get in the past - but I believe this is for their own good. \r\nHowever when I first ran into this I didn't notice that BitVector.size() would become wrong as result of this - nice catch Mike!\r\n\r\nI think however that the test Mike added does not expose the docs out of order bug - I tried this test without the fix and it only fail on the \"gotException assert\" - if you comment this assert the test pass. \r\n\r\nThe following test would expose the out-of-order bug - it would fail with out-of-order before the fix, and would succeed without it. \r\n\r\n  public void testOutOfOrder () throws IOException {\r\n    String tempDir = System.getProperty(\"java.io.tmpdir\");\r\n    if (tempDir == null) {\r\n      throw new IOException(\"java.io.tmpdir undefined, cannot run test: \"+getName());\r\n    }\r\n    \r\n    File indexDir = new File(tempDir, \"lucenetestindexTemp\");\r\n    Directory dir = FSDirectory.getDirectory(indexDir, true);\r\n\r\n    boolean create = true;\r\n    int numDocs = 0;\r\n    int maxDoc = 0;\r\n    while (numDocs < 100) {\r\n      IndexWriter iw = new IndexWriter(dir,anlzr,create);\r\n      create = false;\r\n      iw.setUseCompoundFile(false);\r\n      for (int i=0; i<2; i++) {\r\n        Document d = new Document();\r\n        d.add(new Field(\"body\",\"body\"+i,Store.NO,Index.UN_TOKENIZED));\r\n        iw.addDocument(d);\r\n      }\r\n      iw.optimize();\r\n      iw.close();\r\n      IndexReader ir = IndexReader.open(dir);\r\n      numDocs = ir.numDocs();\r\n      maxDoc = ir.maxDoc();\r\n      assertEquals(numDocs,maxDoc);\r\n      for (int i=7; i >=-1; i--) {\r\n        try {\r\n          ir.deleteDocument(maxDoc+i);\r\n        } catch (ArrayIndexOutOfBoundsException e) {  \r\n        }\r\n      }\r\n      ir.close();\r\n    }\r\n  }\r\n\r\nMike, do you agree?\r\n\r\n(3) maxDoc() computation in SegmentReader is based (on some paths) in RandomAccessFile.length(). IIRC I saw cases (in previous project) where File.length() or RAF.length() (not sure which of the two) did not always reflect real length, if the system was very busy IO wise, unless FD.sync() was called (with performance hit). \r\n\r\nThis post seems relevant - RAF.length over 2GB in NFS - http://forum.java.sun.com/thread.jspa?threadID=708670&messageID=4103657 \r\n\r\nNot sure if this can be the case here but at least we can discuss whether it is better to always store the length.\r\n\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=doronc","name":"doronc","emailAddress":"cdoronc at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=doronc&avatarId=12738","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=doronc&avatarId=12738","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=doronc&avatarId=12738","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=doronc&avatarId=12738"},"displayName":"Doron Cohen","active":true},"created":"2007-01-09T01:27:24.171+0000","updated":"2007-01-09T01:27:24.171+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463202","id":"12463202","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"Hi Michael,\r\n\r\nThis is awesome, I have prepared a patched 1.9.1: http://jira.atlassian.com/secure/attachment/19390/lucene-core-1.9.1-atlassian-patched-2007-01-09.jar\r\n\r\nUnfortunately we don't have a repeatable test for this so we will have to distribute to afflicted customers and - well, pray I guess. We have been seeing this sporadically in our main JIRA instance http://jira.atlassian.com so we will hopefully not observe it now.\r\n\r\nWe do only use the deleteDocuments(Term) method, so we are not sure whether this will truly fix our problem, but we note that that method calls deleteDocument(int) based on the TermDocs returned for the Term - and maybe they can be incorrect???\r\n\r\nOut of interest, apart from changing from 1.4.3 to 1.9.1, in the JIRA 3.7 release we changed our default merge factor to 4 from 10. We hadn't seen this problem before, and suddenly we have had a reasonable number of occurrences. ","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-09T06:38:56.802+0000","updated":"2007-01-09T06:38:56.802+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463203","id":"12463203","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"Alas, this doesn't appear to be the problem. We are still getting it, but we do at least have a little more info. We added the doc and lastDoc to the IllegalArgEx and we are getting very strange numbers:\r\n\r\njava.lang.IllegalStateException: docs out of order (-1764 < 0)\r\n        at org.apache.lucene.index.SegmentMerger.appendPostings([Lorg/apache/lucene/index/SegmentMergeInfo;I)I(SegmentMerger.java:335)\r\n        at org.apache.lucene.index.SegmentMerger.mergeTermInfo([Lorg/apache/lucene/index/SegmentMergeInfo;I)V(SegmentMerger.java:298)\r\n        at org.apache.lucene.index.SegmentMerger.mergeTermInfos()V(SegmentMerger.java:272) \r\n        at org.apache.lucene.index.SegmentMerger.mergeTerms()V(SegmentMerger.java:236)\r\n        at org.apache.lucene.index.SegmentMerger.merge()I(SegmentMerger.java:89)\r\n        at org.apache.lucene.index.IndexWriter.mergeSegments(II)V(IndexWriter.java:681)\r\n        at org.apache.lucene.index.IndexWriter.mergeSegments(I)V(IndexWriter.java:658)\r\n        at org.apache.lucene.index.IndexWriter.maybeMergeSegments()V(IndexWriter.java:646)\r\n        at org.apache.lucene.index.IndexWriter.addDocument(Lorg/apache/lucene/document/Document;Lorg/apache/lucene/analysis/Analyzer;)V(IndexWriter.java:453) \r\n        at org.apache.lucene.index.IndexWriter.addDocument(Lorg/apache/lucene/document/Document;)V(IndexWriter.java:436)\r\n\r\nwhere doc = -1764 and lastDoc is zero","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-09T06:41:40.371+0000","updated":"2007-01-09T06:41:40.371+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463243","id":"12463243","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"Jed, thanks for testing the fix!\r\n\r\n> Alas, this doesn't appear to be the problem. We are still getting\r\n> it, but we do at least have a little more info. We added the doc and\r\n> lastDoc to the IllegalArgEx and we are getting very strange numbers:\r\n>\r\n> java.lang.IllegalStateException: docs out of order (-1764 < 0)\r\n> ...\r\n>\r\n> where doc = -1764 and lastDoc is zero\r\n\r\nOK, so we've definitely got something else at play here (bummer!). That\r\n(negative doc number) is very strange.  I will keep looking a this.  I\r\nwill prepare a patch on 1.9.1 that adds some more instrumentation so\r\nwe can get more details when you hit this ...\r\n\r\n> We do only use the deleteDocuments(Term) method, so we are not sure\r\n> whether this will truly fix our problem, but we note that that\r\n> method calls deleteDocument(int) based on the TermDocs returned for\r\n> the Term - and maybe they can be incorrect???\r\n\r\nRight, but I had thought the docNum's coming in from this path would\r\nbe correct.  It sounds like we have another issue at play here that\r\ncan somehow get even these doc numbers messed up.\r\n\r\n> Out of interest, apart from changing from 1.4.3 to 1.9.1, in the\r\n> JIRA 3.7 release we changed our default merge factor to 4 from\r\n> 10. We hadn't seen this problem before, and suddenly we have had a\r\n> reasonable number of occurrences.\r\n\r\nInteresting.  Maybe try changing back to 4 and see if it suppresses\r\nthe bug?  Might give us [desperately needed] more data to cling to\r\nhere!  On the 1.4.3 -> 1.9.1 change, some of the cases above were even\r\npre-1.4.x (though they could have been from yet another root cause or\r\nmaybe filesystem) so it's hard to draw hard conclusions on this\r\nfront.\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-09T11:18:16.051+0000","updated":"2007-01-09T11:18:16.051+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463247","id":"12463247","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"\r\nDoron,\r\n\r\n> (1) the sequence of ops brought by Jason is wrong:\r\n>   ...\r\n>\r\n> Problem here is that the docIDs found in (b) may be altered in step\r\n> (d) and so step (f) would delete the wrong docs. In particular, it\r\n> might attempt to delete ids that are out of the range. This might\r\n> expose exactly the BitVector problem, and would explain the whole\r\n> thing, but I too cannot see how it explains the delete-by-term case.\r\n\r\nRight, the case I fixed only happens when the Lucene\r\ndeleteDocument(int docNum) is [slightly] mis-used.  Ie if you are\r\n\"playing by the rules\" you would never have hit this bug.  And this\r\nparticular use case is indeed incorrect: doc numbers are only valid to\r\nthe one reader that you got them from.\r\n\r\n> I think however that the test Mike added does not expose the docs\r\n> out of order bug - I tried this test without the fix and it only\r\n> fail on the \"gotException assert\" - if you comment this assert the\r\n> test pass.\r\n\r\nHuh, I see my test case (in IndexReader) indeed hitting the original\r\n\"docs out of order\" exception.  If I take the current trunk and\r\ncomment out the (one line) bounds check in BitVector.set and run that\r\ntest, it hits the \"docs out of order\" exception.\r\n\r\nAre you sure you updated the change (to tighten the check to a <= from\r\na <) to index/SegmentMerger.java?  Because, I did indeed find that the\r\ntest failed to fail when I first wrote it (but should have).  So in\r\ndigging why it didn't fail as expected, I found that the check for\r\n\"docs out of order\" missed the boundary case of the same doc number\r\ntwice in a row.  Once I fixed that, the test failed as expected.\r\n\r\n> (3) maxDoc() computation in SegmentReader is based (on some paths)\r\n> in RandomAccessFile.length(). IIRC I saw cases (in previous project)\r\n> where File.length() or RAF.length() (not sure which of the two) did\r\n> not always reflect real length, if the system was very busy IO wise,\r\n> unless FD.sync() was called (with performance hit).\r\n\r\nYes I saw this too.  From the follow-on discussion it sounds like we\r\nhaven't found a specific known JVM bug here.  Still, it does make me\r\nnervous that we rely on file length to derive maxDoc.\r\n\r\nIn general I think we should rely on as little as possible from the\r\nfile system (there are so many cross platform issues/differences) and\r\ninstead explicitly store things like maxDoc into the index.  I will\r\nopen a separate Jira issue to track this.  Also I will record this\r\npath in the instrumentation patch for 1.9.1 just to see if we are\r\nactually hitting something here (I think unlikely but possible).\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-09T11:32:04.270+0000","updated":"2007-01-09T11:32:04.270+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463249","id":"12463249","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"OK, I created LUCENE-767 for the \"maxDoc should be explicitly stored in the index\" issue.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-09T11:40:58.334+0000","updated":"2007-01-09T11:40:58.334+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463294","id":"12463294","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"\r\nJed, one question: when you tested the fix, you fully rebuilt your\r\nindex from scratch, right?  Just want to verify that.  You have to\r\nre-index because once the index is corrupted it will eventually hit\r\nthe \"docs out of order\" exception even if you fix the original cause.\r\n\r\nOK I've prepared a patch off 1.9.1 (just attached it).  The patch\r\npasses all unit tests on 1.9.1.\r\n\r\nIt has the changes I committed to the trunk yesterday, plus\r\ninstrumentation (messages printed to a PrintStream) to catch places\r\nwhere doc numbers are not correct.\r\n\r\nAll messages I added print to a newly added infoStream static member\r\nof SegmentMerger.  You can do SegmentMerger.setInfoStream(...) to\r\nchange it (it defaults to System.err).\r\n\r\nJed if you could get the error to re-occur with this patch and then\r\npost the resulting messages, that would be great.  Hopefully it gives\r\nus enough information to find the source here or at least to have\r\nanother iteration with yet more instrumentation.  Thanks!\r\n\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-09T14:50:21.430+0000","updated":"2007-01-09T14:50:21.430+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463440","id":"12463440","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"Hi Michael,\r\n\r\nThanks for the patch, applied and recreated. Attached is the log.\r\n\r\nTo be explicit, we are recreating the index via the IndexWriter ctor with the create flag set and then completely rebuilding the index. We are not completely deleting the entire directory. There ARE old index files (_*.cfs & _*.del) in the directory with updated timestamps that are months old. If I completely recreate the directory the problem does go away. This is a fairly trivial \"fix\", but we are still investigating as we want to know if this is indeed the problem, how we have come to make it prevalent, and what the root cause is.\r\n\r\nThanks for all the help everyone.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-10T00:57:48.451+0000","updated":"2007-01-10T00:57:48.451+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463470","id":"12463470","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"BTW. We have looked at all the open files referenced by the VM when the indexing errors occur, and there does not seem to be any reference to the old index segment files, so I am not sure how those files are influencing this problem.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-10T06:07:58.268+0000","updated":"2007-01-10T06:07:58.268+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463483","id":"12463483","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=doronc","name":"doronc","emailAddress":"cdoronc at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=doronc&avatarId=12738","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=doronc&avatarId=12738","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=doronc&avatarId=12738","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=doronc&avatarId=12738"},"displayName":"Doron Cohen","active":true},"body":"Jed, is it possible that when re-creating the index, while IndexWriter is constructed with create=true, FSDirectory is opened with create=false?\r\nI suspect so, because otherwise, old  .del files would have been deleted. \r\nIf indeed so, newly created segments, which have same names as segments in previous (bad) runs, when opened, would read the (bad) old .del file. \r\nThis would possibly expose the bug fixed by Michael. \r\nI may be over speculating here, but if this is the case, it can also explain why changing the merge factor from 4 to 10 exposed the problem. \r\n\r\nIn fact, let me speculate even further - if indeed when creating the index from scratch, the FSDirectory is (mistakenly) opened with create=false, as long as you always repeated the same sequencing of adding and deleting docs, you were likely to almost not suffer from this mistake, because segments created with same names as (old) .del files simply see docs as deleted before the docs are actually deleted by the program. The search behaves wrongly, not finding these docs before they are actually deleted, but no exception is thrown when adding docs. However once the merge factor was changed from 4 to 10, the matching between old .del files and new segments (with same names) was broken, and the out-of-order exception appeared. \r\n\r\n...and if this is not the case, we would need to look for something else...","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=doronc","name":"doronc","emailAddress":"cdoronc at gmail dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?ownerId=doronc&avatarId=12738","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&ownerId=doronc&avatarId=12738","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&ownerId=doronc&avatarId=12738","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&ownerId=doronc&avatarId=12738"},"displayName":"Doron Cohen","active":true},"created":"2007-01-10T07:36:28.738+0000","updated":"2007-01-10T07:36:28.738+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463524","id":"12463524","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"OK from that indexing-failure.log (thanks Jed!) I can see that indeed\r\nthere are segments whose maxDoc() is much smaller than\r\ndeleteDocs.count().  This then leads to negative doc numbers on\r\nmerging these segments.\r\n\r\nJed when you say \"there are old files (_*.cfs & _*.del) in this\r\ndirectory with updated timestamps that are months old\" what do you\r\nmean by \"with updated timestamps\"?  Which timestamp is months old and\r\nwhich one is updated?\r\n\r\nOK, assuming Jed you are indeed sending \"create=false\" when creating\r\nthe Directory and then passing that directory to IndexWriter with\r\ncreate=true, I think we now have this case fully explained (thanks\r\nDoron): your old _*.del files are being incorrectly opened & re-used\r\nby Lucene, when they should not be.\r\n\r\nLucene (all released versions but not the trunk version, see below)\r\ndoes a simple fileExists(\"_XXX.del\") call to determine if a segment\r\nXXX has deletes.\r\n\r\nBut when that _XXX.del is a leftover from a previous index, it very\r\nlikely doesn't \"match\" the newly created _XXX segment.  (Especially if\r\nmerge factor has changed but also if order of operations has changed,\r\nwhich I would expect in this use case).\r\n\r\nIf that file exists, Lucene assumes it's for this segment and so opens\r\nit and uses it.  If it happens that this _XXX.del file has more\r\ndocuments in it than the newly created _XXX.cfs segment, then negative\r\ndoc numbers will result (and then later cause the \"docs out of order\"\r\nexception).  If it happens that the _XXX.del file has fewer documents\r\nthan the newly created _XXX.cfs segment then you'll hit\r\nArrayIndexOutOfBounds exceptions in calls to isDeleted(...).  If they\r\nare exactly equal then you'd randomly see some of your docs got\r\ndeleted.\r\n\r\nNote that the trunk version of Lucene has already fixed this bug (as\r\npart of lockless commits):\r\n\r\n  * Whether a segment has deletions or not is now explictly stored in\r\n    the segments file rather than relying on a \"fileExists(...)\" call.\r\n    So, if an old _XXX.del existed in the filesystem, the newly\r\n    created _XXX segment would not open it.\r\n\r\n  * Furthermore, the trunk version of Lucene uses a new\r\n    IndexFileDelter class to remove any unreferenced index files.\r\n    This means it would have removed these old _*.cfs and _*.del files\r\n    even in the case where a directory was created with \"create=false\"\r\n    and the IndexWriter was created with \"create=true\".\r\n\r\nTo summarize:\r\n\r\n  * There was one case where if you gave slightly illegal doc numbers\r\n    (within 7 of the actual maxDoc) Lucene may silently accept the\r\n    call but would corrupt your index only to be seen later as an\r\n    \"docs out of order\" IllegalStateException when the segment is\r\n    merged.  This was just a missing boundary case check.  This case\r\n    is now fixed in the trunk (you get an\r\n    ArrayIndexOutOfBoundsException if doc number is too large).\r\n\r\n  * There is also another case, that only happens if you have old\r\n    _*.del files leftover from a previous index while re-creating a\r\n    new index.\r\n\r\n    The workaround is simple here: always open the Directory with\r\n    create=true (or, remove the directory contents yourself before\r\n    hand).  (IndexWriter does this if you give it a String or File\r\n    with create=true).\r\n\r\n    This is really a bug in Lucene, but given that it's already fixed\r\n    in the trunk, and the workaround is simple, I'm inclined to not\r\n    fix it in prior releases and instead publicize the issue (I will\r\n    do so on java-user).\r\n\r\n    But, I will commit two additional IllegalStateException checks to\r\n    the trunk when a segment is first initialized: 1) check that the\r\n    two different sources of \"maxDoc\" (fieldsReader.size() and\r\n    si.docCount) are the same, and 2) check that the number of pending\r\n    deletions does not exceed maxDoc().  When an index has\r\n    inconsistency I think the earlier it's detected the better.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-10T10:48:05.329+0000","updated":"2007-01-10T10:48:05.329+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463674","id":"12463674","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"> BTW. We have looked at all the open files referenced by the VM when\r\n> the indexing errors occur, and there does not seem to be any reference\r\n> to the old index segment files, so I am not sure how those files are\r\n> influencing this problem.\r\n\r\nJed just to answer this question: the _XXX.del files are opened very\r\nbriefly because the contents of this file are loaded / cached in\r\nmemory, and the the file handle is closed.  I don't think the _XXX.cfs\r\nfiles are affecting this issue (are not opened).\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-10T19:18:57.631+0000","updated":"2007-01-10T19:18:57.631+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463781","id":"12463781","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"body":"Michael, Doron, you guys are legends!\r\n\r\nIndeed the problem is using only the IndexWriter with create true to recreate the directory. Creating a new Directory with create true does fix the problem. The javadoc for this constructor is fairly explicit that it should recreate the index for you (no caveat), so I would consider that a bug, but - given that head fixes it - not one that requires any action.\r\n\r\nThanks guys for the prompt attention, excellent and thorough analysis.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=jedws","name":"jedws","emailAddress":"jed at atlassian dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Jed Wesley-Smith","active":true},"created":"2007-01-11T01:18:32.607+0000","updated":"2007-01-11T01:18:32.607+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463872","id":"12463872","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"Phew!  I'm glad we finally got to the bottom of this one.\r\n\r\nThank you for your persistent and fast testing iterations, Jed; this\r\nissue has been open for far too long!\r\n\r\nI will send a summary email to java-user and resolve this issue,\r\nfinally.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-11T11:46:20.202+0000","updated":"2007-01-11T11:46:20.202+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463875","id":"12463875","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"\r\nActually, this reminds me that, as of lockless commits, there is one\r\nimportant tradeoff on which \"create=true\" to use (the case on windows\r\nwhere you want to re-create the index but readers are currently using\r\nit).  I will call out this difference in the javadocs.\r\n\r\nAlthough, why do we even have a \"create\" parameter in the directory?\r\nI think it's confusing (and dangerous, pre-trunk, due to this issue)\r\nto have two ways of doing the same thing?\r\n\r\nLogically, I don't think a Directory should take the responsibility of\r\ndeleting old files (including old lock files).  It should be a clean\r\ninterface for doing so, but I think the IndexWriter alone should be\r\nthe class that deletes files from the directory.\r\n\r\nWith lockless commits this has become an important difference, ie, the\r\nnew IndexFileDeleter class (used by IndexWriter) handles re-trying\r\nfiles that are in-use (on Windows) whereas FSDirectory will throw an\r\nexception if create=true and there are index files are in use.\r\n\r\nI think we should deprecate the \"create\" argument to\r\nFSDirectory.getDirectory(*) and leave only the create argument in\r\nIndexWriter's constructors.  Am I missing something?  Is there are a\r\nreason not to do this?\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-11T12:09:01.221+0000","updated":"2007-01-11T12:09:01.221+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12463877","id":"12463877","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"Resolving this now, finally (I'll move the two \"create\" arguments\r\ndiscussion to a separate issue if we decide to go forward with that):\r\n\r\n  * Fixed (in the trunk) to catch boundary cases of incorrect docNum's\r\n    to deleteDocuments.\r\n\r\n  * Fixed (in the trunk) to do earlier \"IllegalState\" checks to catch\r\n    index corruption sooner.  Also fixed the existing IllegalState\r\n    check to catch missing boundary cases.\r\n\r\n  * The re-using of old _XXX.del files is already fixed with lockless\r\n    commits (in trunk).  This remains open for past releases, but the\r\n    workaround is simple and I've now publicized this on java-user.\r\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-01-11T12:14:18.277+0000","updated":"2007-01-11T12:14:18.277+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12314290/comment/12476252","id":"12476252","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"body":"Closing all issues that were resolved for 2.1.","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=mikemccand","name":"mikemccand","emailAddress":"lucene at mikemccandless dot com","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Michael McCandless","active":true},"created":"2007-02-27T18:10:32.036+0000","updated":"2007-02-27T18:10:32.036+0000"}]},"votes":{"self":"https://issues.apache.org/jira/rest/api/2/issue/LUCENE-140/votes","votes":5,"hasVoted":false},"worklog":{"startAt":0,"maxResults":20,"total":0,"worklogs":[]},"customfield_12311820":"0|i0550f:"}}